gadget/update.go

```diff
@@ -665,6 +668,139 @@ func DiskTraitsFromDeviceAndValidate(expLayout *LaidOutVolume, dev string, opts
+// volume containing the system-boot role, when we cannot load an existing
+// traits object from disk-mapping.json. It is meant to be used only with all
+// UC16/UC18 installs as well as UC20 installs from before we started writing
+// disk-mapping.json during install mode.
+func buildNewVolumeToDeviceMapping(old GadgetData, laidOutVols map[string]*LaidOutVolume, preUC20 bool) (map[string]DiskVolumeDeviceTraits, error) {
+	var systemBootVolume string
+
+	// we need to pick the volume, since updates for this setup are best
+	// effort and mainly focused on the volume with system-* roles on it, we
+	// need to pick the volume with that role
+volumeLoop:
+	for volName, vol := range old.Info.Volumes {
+		for _, structure := range vol.Structure {
+			if structure.Role == SystemBoot {
+				// this is the volume
+				systemBootVolume = volName
+				break volumeLoop
+			}
+		}
+	}
+	if systemBootVolume == "" {
+		// didn't find system-boot anywhere somehow
+		if preUC20 {
+			logger.Noticef("WARNING: could not identify disk for gadget asset update of volume %s: unable to find any volume with system-boot role on it", systemBootVolume)
+			return nil, errSkipUpdateProceedRefresh
+		}
+		// shouldn't be possible on UC20
+		return nil, fmt.Errorf("cannot find any volume with system-boot, gadget is broken")
+	}
+
+	laidOutVol := laidOutVols[systemBootVolume]
+
+	// search for matching devices that correspond to the volume we laid out
+	dev := ""
+	for _, vs := range laidOutVol.LaidOutStructure {
+		// here it is okay that we require there to be either a partition label
+		// or a filesystem label since we require there to be a system-boot role
+		// on this volume which by definition must have a filesystem
+		structureDevice, err := FindDeviceForStructure(&vs)
+		if err == ErrDeviceNotFound {
+			continue
+		}
+		if err != nil {
+			// TODO: should this be a fatal error?
+			return nil, err
+		}
+
+		// we found a device for this structure, get the parent disk
+		// and save that as the device for this volume
+		disk, err := disks.DiskFromPartitionDeviceNode(structureDevice)
+		if err != nil {
+			// TODO: should we keep looping instead and try again with
+			// another structure? it probably wouldn't work because we found
+			// something on disk with the same name as something from the
+			// gadget.yaml, but then we failed to make a disk from that
+			// partition which suggests something is inconsistent with the
+			// state of the disk/udev database
+			return nil, err
+		}
+
+		dev = disk.KernelDeviceNode()
+		break
+	}
+
+	if dev == "" {
+		// couldn't find a disk at all, pre-UC20 we just warn about this
+		// but let the update continue
+		if preUC20 {
+			logger.Noticef("WARNING: could not identify disk for gadget asset update of volume %s", systemBootVolume)
+			return nil, errSkipUpdateProceedRefresh
+		}
+		// fatal error on UC20+
+		return nil, fmt.Errorf("could not identify disk for gadget asset update of volume %s", systemBootVolume)
```

Comment:
same about cannot, sorry I should made punctual comments, instead of a general one

Reply:
hopefully got them all this time